File: src/SimpleLevelEditor.Formats/Types/EntityConfig/EntityDescriptor.fs
namespace SimpleLevelEditor.Formats.Types.EntityConfig

type EntityDescriptor =
    { Name: string
      Shape: EntityShapeDescriptor
      Properties: EntityPropertyDescriptor list }

    member this.DeepCopy() =
        { this with
            Name = this.Name
            Shape = this.Shape.DeepCopy()
            Properties = this.Properties |> List.map (_.DeepCopy()) }

File: src/SimpleLevelEditor.Formats.EntityConfig/EntityDescriptor.cs
using System.Diagnostics.CodeAnalysis;

namespace SimpleLevelEditor.Formats.EntityConfig;

public record EntityDescriptor
{
	public required string Name { get; init; }

	public required EntityShapeDescriptor Shape { get; init; }

	public required List<EntityPropertyDescriptor> Properties { get; init; }

	[SetsRequiredMembers]
	public EntityDescriptor(string name, EntityShapeDescriptor shape, List<EntityPropertyDescriptor> properties)
	{
		Name = name;
		Shape = shape;
		Properties = properties;
	}

	public EntityDescriptor DeepCopy()
	{
		return this with
		{
			Shape = Shape.DeepCopy(),
			Properties = Properties.Select(p => p.DeepCopy()).ToList(),
		};
	}
}
